lib/utils/DevServerPlugin.js

```diff
@@ -0,0 +1,209 @@
+      options.client.path !== '/ws'
+    ) {
+      path = `&path=${options.client.path}`;
+    }
+    /** @type {string} */
+    const port =
+      options.client && options.client.port
+        ? `&port=${options.client.port}`
+        : '';
+    /** @type {string} */
+    const clientEntry = `${require.resolve(
+      '../../client/default/'
+    )}?${domain}${host}${path}${port}`;
+
+    /** @type {(string[] | string)} */
+    let hotEntry;
+
+    if (options.hot === 'only') {
+      hotEntry = require.resolve('webpack/hot/only-dev-server');
+    } else if (options.hot) {
+      hotEntry = require.resolve('webpack/hot/dev-server');
+    }
+    /**
+     * prependEntry Method for webpack 4
+     * @param {Entry} originalEntry
+     * @param {Entry} additionalEntries
+     * @returns {Entry}
+     */
+    const prependEntry = (originalEntry, additionalEntries) => {
+      if (typeof originalEntry === 'function') {
+        return () =>
+          Promise.resolve(originalEntry()).then((entry) =>
+            prependEntry(entry, additionalEntries)
+          );
+      }
+
+      if (typeof originalEntry === 'object' && !Array.isArray(originalEntry)) {
+        /** @type {Object<string,string>} */
+        const clone = {};
+
+        Object.keys(originalEntry).forEach((key) => {
+          // entry[key] should be a string here
+          const entryDescription = originalEntry[key];
+          clone[key] = prependEntry(entryDescription, additionalEntries);
+        });
+
+        return clone;
+      }
+
+      // in this case, entry is a string or an array.
+      // make sure that we do not add duplicates.
+      /** @type {Entry} */
+      const entriesClone = additionalEntries.slice(0);
+      [].concat(originalEntry).forEach((newEntry) => {
+        if (!entriesClone.includes(newEntry)) {
+          entriesClone.push(newEntry);
+        }
+      });
+      return entriesClone;
+    };
+
+    /**
+     *
+     * Description of the option for checkInject method
+     * @typedef {Function} checkInjectOptionsParam
+     * @param {Object} _config - compilerConfig
+     * @return {Boolean}
+     */
+
+    /**
+     *
+     * @param {Boolean | checkInjectOptionsParam} option - inject(Hot|Client) it is Boolean | fn => Boolean
+     * @param {Object} _config
+     * @param {Boolean} defaultValue
+     * @return {Boolean}
+     */
+    // eslint-disable-next-line no-shadow
+    const checkInject = (option, _config, defaultValue) => {
+      if (typeof option === 'boolean') return option;
+      if (typeof option === 'function') return option(_config);
+      return defaultValue;
+    };
+
+    const compilerOptions = compiler.options;
+    compilerOptions.plugins = compilerOptions.plugins || [];
+    /** @type {Boolean} */
+    const webTarget = [
+      'web',
+      'webworker',
+      'electron-renderer',
+      'node-webkit',
+      // eslint-disable-next-line no-undefined
+      undefined,
+      null,
+    ].includes(compilerOptions.target);
```

Comment:
I found better way to do it in webpack@5, let's check:
```js
const isWebTarget = compiler.optons.externalsPresets 
  ? compiler.optons.externalsPresets.web
  : // Old logic for webpack@4
```

Reply:
Should we add client entries when `compiler.optons.externalsPresets.web === null`, i.e., targets are mixed, e.g., `['web', 'node']`?

Reply:
No, only when it `web`, if it is not `web` and still in browser (I don't know this case :smile: ) you should use `injectHot: true`